### PR TITLE
add task solution

### PR DESCRIPTION
**Pull request by Koliras**

### CODE REVIEW

File: src/averageAges.js [hunk 1]
```diff
@@ -15,11 +15,14 @@
...
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const men = century
+    ? people.filter(
+      ({ sex, died }) => sex === 'm' && Math.ceil(died / 100) === century)
+    : people.filter(({ sex }) => sex === 'm');
+
+  return men.reduce(
+    (prev, { born, died }) => prev + (died - born), 0)
```
Comment (davinnchii) [inline src/averageAges.js]: I recommend using accumulator, acc or total for a parameter of reduce
[MDN Doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce)

File: src/averageAges.js [hunk 2]
```diff
@@ -37,7 +40,14 @@ function calculateMenAverageAge(people, century) {
...
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = withChildren
+    ? people.filter(({ sex, name }) =>
+      sex === 'f' && people.some(({ mother }) => name === mother))
+    : people.filter(({ sex }) => sex === 'f');
+
+  return women.reduce(
+    (prev, { born, died }) => prev + (died - born), 0)
```
Comment (davinnchii) [inline src/averageAges.js]: > I recommend using accumulator, acc or total for a parameter of reduce
[MDN Doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce)

The same one

Comment (davinnchii) [inline src/averageAges.js]: Actually it is good, but u can use variable for died - born formula or alternatively for 1st and 2nd tasks you can do a function to calculate years
```
function calculateAge(array) {
  return array.reduce((acc, person) => {
    const age = person.died - person.born;
    return acc + age;
  }, 0) / array.length;
}
```

**Review by davinnchii — CHANGES_REQUESTED** [review]
Good job, but I suggest some small changes to improve readability.

File: src/averageAges.js [hunk 3]
```diff
@@ -37,7 +40,14 @@ function calculateMenAverageAge(people, century) {
...
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = withChildren
+    ? people.filter(({ sex, name }) =>
+      sex === 'f' && people.some(({ mother }) => name === mother))
+    : people.filter(({ sex }) => sex === 'f');
+
+  return women.reduce(
+    (total, { born, died }) => total + (died - born), 0)
+    / women.length;
```
Comment (tiserett) [inline src/averageAges.js]: we already use it twice, let's create a function for these calculations

**Review by tiserett — CHANGES_REQUESTED** [review]
LGTM, but let's create a function which I described 

I also want you to think on overall code style, think how you can change it, in array methods

File: src/averageAges.js [hunk 4]
```diff
@@ -14,12 +14,20 @@
...
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const men = century
+    ? people.filter(
+      ({ sex, died }) => sex === 'm' && Math.ceil(died / 100) === century)
+    : people.filter(({ sex }) => sex === 'm');
```
Comment (roman-mirzoian) [inline src/averageAges.js]:
```suggestion
   const filteredPeopleByMan = people.filter(({ sex, died }) => {
     return century ? sex === 'm' && Math.ceil(died / 100) === century) : sex === 'm';
   });
```

File: src/averageAges.js [hunk 5]
```diff
@@ -14,12 +14,20 @@
...
+
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const men = century
+    ? people.filter(
+      ({ sex, died }) => sex === 'm' && Math.ceil(died / 100) === century)
```
Comment (roman-mirzoian) [inline src/averageAges.js]: It would be cool to put all the "magic values" in separate constants.

File: src/averageAges.js [hunk 6]
```diff
@@ -36,8 +44,14 @@ function calculateMenAverageAge(people, century) {
  *
  * @return {number}
  */
+
 function calculateWomenAverageAge(people, withChildren) {
```
Comment (roman-mirzoian) [inline src/averageAges.js:52]: The requirements for this function are the same as for the previous one.

File: src/averageAges.js [hunk 7]
```diff
@@ -55,10 +69,16 @@ function calculateWomenAverageAge(people, withChildren) {
...
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const children = onlyWithSon
+    ? people.filter(({ sex, mother }) =>
+      (people.some(({ name }) => mother === name) && sex === 'm'))
+    : people.filter(({ mother }) =>
+      people.some(({ name }) => mother === name));
```
Comment (roman-mirzoian) [inline src/averageAges.js]:
```suggestion
    const filteredPeopleByMother = people.filter(({ sex, mother: motherName }) => {
      const mother = people.find(({ name }) => motherName === name);
      return onlyWithSon ? mother && sex === 'm' : mother;
    });
```

**Review by roman-mirzoian — CHANGES_REQUESTED** [review]
Good job!
But you should not try to write the entire code "in one line", it makes it unreadable and if you need to change something in the future, it will be very difficult.
Try to break large parts of the code into smaller ones, choosing appropriate variable names that will explain what is happening in the code.
This applies to all code (all functions).

File: src/averageAges.js [hunk 8]
```diff
@@ -14,12 +14,22 @@
  *
  * @return {number}
  */
+
+function averageAges(people) {
```
Comment (Moroz-Dmytro) [inline src/averageAges.js]: image

File: src/averageAges.js [hunk 9]
```diff
@@ -14,12 +14,22 @@
...
+  return people.reduce(
+    (total, { born, died }) => total + (died - born), 0)
+    / people.length;
```
Comment (Moroz-Dmytro) [inline src/averageAges.js]:
```suggestion
  return people.reduce((total, { born, died }) => (
    total + (died - born)
  ), 0) / people.length;
```

File: src/averageAges.js [hunk 10]
```diff
@@ -14,12 +14,22 @@
...
+}
+
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const CENTURY = 100;
```
Comment (Moroz-Dmytro) [inline src/averageAges.js]: Move it outside the function, to the beginning of file

File: src/averageAges.js [hunk 11]
```diff
@@ -14,12 +14,22 @@
...
+  const filteredPeopleByMan = people.filter(({ sex, died }) => {
+    return century
+      ? sex === 'm' && Math.ceil(died / CENTURY) === century
```
Comment (Moroz-Dmytro) [inline src/averageAges.js]: `m` `f`  - magic values. Let's create constant values for them

File: src/averageAges.js [hunk 12]
```diff
@@ -55,10 +72,17 @@ function calculateWomenAverageAge(people, withChildren) {
...
+    ({ sex, mother: motherName }) => {
+      const mother = people.find(({ name }) => motherName === name);
+
+      return onlyWithSon ? mother && sex === 'm' : mother;
+    });
+
+  return filteredPeopleByMother.reduce(
+    (total, { born, mother }) =>
+      total + (born - people.find(({ name }) => name === mother).born), 0)
+    / filteredPeopleByMother.length;
```
Comment (Moroz-Dmytro) [inline src/averageAges.js]: Let's simplify this. Create variable for `filteredPeopleByMother.reduce(...)`.
Create variable for `people.find(({ name }) => name === mother).born` inside `reduce`

**Review by Moroz-Dmytro — CHANGES_REQUESTED** [review]
GJ 👍 
But small comments

File: src/averageAges.js [hunk 13]
```diff
@@ -14,12 +14,24 @@
...
  * @return {number}
  */
+const CENTURY = 100;
+const MALE = 'm';
+const FEMALE = 'f';
+
+function calculateAverageAges(people) {
+  return people.reduce((total, { born, died }) =>
+    total + (died - born
+    ), 0) / people.length;
```
Comment (Moroz-Dmytro) [inline src/averageAges.js:24]:
```suggestion
  return people.reduce((total, { born, died }) => (
    total + (died - born)
  ), 0) / people.length;
```

**Review by Moroz-Dmytro — APPROVED** [review]
GJ 👍

**Pull request closed without merging**